### PR TITLE
Include host scope for container joins based on IP.

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #520 - can explain false edges between containers on different hosts, but the bug has been seen between containers on the same host, and this shouldn't affect that.

### CODE REVIEW

File: probe/docker/container.go [hunk 1]
```diff
@@ -219,20 +220,28 @@ func (c *container) ports(localAddrs []net.IP) string {
 	return strings.Join(ports, ", ")
 }
 
-func (c *container) GetNode(localAddrs []net.IP) report.Node {
+func (c *container) GetNode(hostID string, localAddrs []net.IP) report.Node {
 	c.RLock()
 	defer c.RUnlock()
 
+	ips := append(c.container.NetworkSettings.SecondaryIPAddresses,
+		c.container.NetworkSettings.IPAddress)
```
Comment (peterbourgon) [inline probe/docker/container.go]: Either

```go
ips := append(c.container.NetworkSettings.SecondaryIPAddresses, c.container.NetworkSettings.IPAddress)
```

or

```go
ips := append(
	c.container.NetworkSettings.SecondaryIPAddresses,
	c.container.NetworkSettings.IPAddress,
)
```

File: probe/docker/container.go [hunk 2]
```diff
@@ -219,20 +220,28 @@ func (c *container) ports(localAddrs []net.IP) string {
...
+	// Treat all Docker IPs as local scoped.
+	ipsWithScopes := []string{}
+	for _, ip := range ips {
+		ipsWithScopes = append(ipsWithScopes, fmt.Sprintf("%s:%s", hostID, ip))
```
Comment (peterbourgon) [inline probe/docker/container.go]: We have ID constructors for operations like this; can you use one?

File: probe/docker/container.go [hunk 3]
```diff
@@ -268,3 +277,9 @@ func (c *container) GetNode(localAddrs []net.IP) report.Node {
 func ExtractContainerIPs(nmd report.Node) []string {
 	return strings.Fields(nmd.Metadata[ContainerIPs])
 }
+
+// ExtractContainerIPsWithScopes returns the list of container IPs, prepended
+// with scopes, given a Node from the Container topology.
+func ExtractContainerIPsWithScopes(nmd report.Node) []string {
+	return strings.Fields(nmd.Metadata[ContainerIPsWithScopes])
+}
```
Comment (peterbourgon) [inline probe/docker/container.go]: Likewise above, we have ID parsers instead of falling back to raw strings.Fields...

File: render/mapping.go [hunk 4]
```diff
@@ -283,23 +283,22 @@ func MapEndpoint2IP(m RenderableNode, local report.Networks) RenderableNodes {
...
-	// Emit addr:port nodes as well, so connections from the internet to containers
-	// via port mapping also works.
-	port, ok := m.Metadata[endpoint.Port]
-	if ok {
-		id := fmt.Sprintf("%s:%s", addr, port)
-		result[id] = NewRenderableNodeWith(id, "", "", "", m)
+	// Emit 2 nodes: scope:addr and scope:addr:port, so connections from the
+	// internet to containers via port mapping also works.
+	id := fmt.Sprintf("%s:%s", scope, addr)
+	idWithPort := fmt.Sprintf("%s:%s:%s", scope, addr, port)
```
Comment (peterbourgon) [inline render/mapping.go]: Likewise above, can you use constructors?

File: render/mapping.go [hunk 5]
```diff
@@ -283,23 +283,22 @@ func MapEndpoint2IP(m RenderableNode, local report.Networks) RenderableNodes {
...
+	return RenderableNodes{
+		id:         NewRenderableNodeWith(id, "", "", "", m),
+		idWithPort: NewRenderableNodeWith(idWithPort, "", "", "", m),
```
Comment (peterbourgon) [inline render/mapping.go]: Why do you need both? Isn't a scoped-X strictly more useful than, and usable in the place of, X alone?

File: render/mapping.go [hunk 6]
```diff
@@ -309,20 +308,20 @@ var portMappingMatch = regexp.MustCompile(`([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.
...
 	}
 
-	// also output all the host:port port mappings
+	// Also output all the host:port port mappings.
+	// In this case, we assume this doesn't need a scope,
+	// as they are for host IPs.
 	for _, mapping := range portMappingMatch.FindAllStringSubmatch(m.Metadata[docker.ContainerPorts], -1) {
 		ip, port := mapping[1], mapping[2]
-		id := fmt.Sprintf("%s:%s", ip, port)
+		id := fmt.Sprintf(":%s:%s", ip, port)
```
Comment (peterbourgon) [inline render/mapping.go]: Likewise above, please use a constructor...

File: render/mapping.go (hunk 5 above)
Comment (tomwilkie) [inline render/mapping.go]: We don't always know what port a container is listening on, and container-to-container communications can be unambiguously identified without ports.  OTOH, connections to the host IPs which have been port mapped to a container can only be unambiguously identified with the port.  So we need both, for two different cases.

Comment (peterbourgon) [inline render/mapping.go]: I understand. Thanks for the explanation!

**Comment by peterbourgon** [comment]
LGTM.

File: render/mapping.go (hunk 5 above)
Comment (tomwilkie) [inline render/mapping.go]: I should probably put it in a comment...

**Comment by tomwilkie** [comment]
Crap; there is actually a good reason to use the id generators - they embed the logic for whether a scope is appended or not, and really are only for use within the probe.

Lets see if can factor this out.

**Comment by tomwilkie** [comment]
```MakeEndpointNodeIDAlwaysWithScope``` seems a little verbose; any better suggestions?

**Comment by peterbourgon** [comment]
MakeScopedEndpointNodeID

or the general form

Make[Scoped]_Topology_NodeID

**Comment by tomwilkie** [comment]
Okay done.  Hopefully this should all pass now.

**Pull request merged**